llvm/lib/IR/AutoUpgrade.cpp

```diff
@@ -5021,6 +5024,36 @@ bool llvm::UpgradeDebugInfo(Module &M) {
   return Modified;
 }
 
+static void upgradeNVVMFnVectorAttr(const StringRef Attr, const char DimC,
+                                    GlobalValue *GV, const Metadata *V) {
+  Function *F = cast<Function>(GV);
+
+  constexpr StringLiteral DefaultValue = "1";
+  StringRef Vect3[3] = {DefaultValue, DefaultValue, DefaultValue};
+  unsigned Length = 0;
+
+  if (F->hasFnAttribute(Attr)) {
+    StringRef S = F->getFnAttribute(Attr).getValueAsString();
+    for (; Length < 3 && !S.empty(); Length++) {
+      auto [Part, Rest] = S.split(',');
+      Vect3[Length] = Part.trim();
+      S = Rest;
+    }
+  }
+
+  const uint64_t VInt = mdconst::extract<ConstantInt>(V)->getZExtValue();
+  const std::string VStr = llvm::utostr(VInt);
```

Comment:
I'd move these down to where we set `Vect3[Dim] = VStr;`. Maybe drop `VStr` and assign directly.

Reply:
Github unhelpfully does not show this comment in the diff, so I almost missed it. :-/ 

Then I'd make Dim3 a SmallVector of SmallString so we can't accidentally shoot ourselves in the foot.

Reply:
The current implementation avoid the need make an additional copy of the strings from the original attr as with StringRefs we can just point to the original data. If we move to SmallString this won't be the case.

Reply:
Fair enough. Just add a comment that we do need a local variable for the reference to point somewhere.
Most likely compiler would do good enough job eliminating unnecessary stuff, and this code is not anywhere near hot path in any case.

Reply:
Comment added